ADR/0027-availability-probe-framework.md

```diff
@@ -0,0 +1,125 @@
+easily visualize and comprehend the stability and availability of deployed systems in
+order to inform and influence future work towards improving the overall system
+reliability.
+  
+Such indication should tell us the overall up time of AppStudio with respect to services
+under the control of AppStudio developers.
+
+AppStudio is defined to be available at a given moment if all of its components are
+reporting to be available at that given moment.
+
+A component is defined to be available at a given moment if all of its availability
+probes are reporting to be available at that given moment.
+
+A convention is required for providing the availability of a probe.
+
+Once this is in place, those indicators can be aggregated in order to report the overall
+availability of AppStudio.
+
+## Decision
+
+Probes' availability will be provided as a Prometheus metric. The metric will contain
+metric labels to allow differentiating between the different probes. When exported out
+of the cluster of origin, additional labels will be attached to the metric to mark the
+cluster in which the metric was generated.
+
+### Delivery Method
+
+The availability Prometheus metric will be computed for each component based on the exit
+status of the latest execution of the CronJobs evaluating the component's availability.
+Developer teams will provide the CronJobs' implementation. The job results will be
+translated a standardized Prometheus metric in a unified fashion (i.e. developers
+will not be required to provide the translation mechanism).
+
+It is up for each team to define what it means for its component(s) to be available.
+Fundamentally, a component should be reported as available as long as it's capable of
+providing the service it aims to provide, and unavailable otherwise.
+
+Each team will define CronJobs that will test the availability of their components.
+The Job started by each CronJob will terminate successfully if the test complete
+successfully, and will terminate with an error in case the test fails.
+
+Kube-state-metrics is a Prometheus exporter generating metrics based on the Kubernetes
+API server. It generates
+[Jobs](https://github.com/kubernetes/kube-state-metrics/blob/main/docs/job-metrics.md) and
+[CronJob](https://github.com/kubernetes/kube-state-metrics/blob/main/docs/cronjob-metrics.md)
+metrics that will be processed using Prometheus recording rules in order to
+generate AppStudio's availability Prometheus metric.
+
+A single set of rules will be defined globally which will apply for all CronJobs.
+
+The resulting Prometheus metric will identify each probe based on its CronJob name and the namespace in which the CronJob was defined:
+
+* The name of the CronJob will be translated to a Prometheus metric label denoting the
+  name of the probe.
+* The namespace in which the CronJob is defined will be translated to a Prometheus
+  metric label denoting the probe's namespace.
```

Comment:
This feels like an implementation detail. Is it relevant to the architectural record? It may be able to be linked together if you word it in a way of the auto-conversion strategy.

Reply:
I removed it now.